packages/app-contracts/src/ABI.tsx

```diff
@@ -5,65 +5,177 @@
 import { I18nProps } from '@polkadot/ui-app/types';
 
 import React from 'react';
-import { InputFile } from '@polkadot/ui-app';
+import styled from 'styled-components';
+import { InputFile, Labelled, Messages } from '@polkadot/ui-app';
 import { ContractAbi } from '@polkadot/types';
 import { u8aToString } from '@polkadot/util';
 
 import translate from './translate';
 
 type Props = I18nProps & {
-  help: React.ReactNode,
+  contractAbi?: ContractAbi,
+  help?: React.ReactNode,
   isError?: boolean,
-  label: React.ReactNode,
-  onChange: (json: string | null, contractAbi: ContractAbi | null) => void
+  isDisabled?: boolean,
+  isRequired?: boolean,
+  label?: React.ReactNode,
+  onChange: (json: string | null, contractAbi: ContractAbi | null) => void,
+  onRemove?: () => void,
+  onRemoved?: () => void,
+  onSelect?: () => void
 };
 
 type State = {
-  abi?: Uint8Array | null,
+  contractAbi?: ContractAbi,
   isAbiValid: boolean,
-  name?: string,
-  placeholder?: React.ReactNode | null
+  isEmpty: boolean,
+  isError: boolean
 };
 
+const Normalize = styled.div`
+  min-height: 4rem;
+`;
+
 class ABI extends React.PureComponent<Props, State> {
   state: State = {
-    isAbiValid: true
+    isAbiValid: false,
+    isEmpty: true,
+    isError: false
   };
 
+  constructor (props: Props) {
+    super(props);
+    if (props.contractAbi) {
+      this.state = {
+        ...this.state,
+        contractAbi: props.contractAbi,
+        isAbiValid: true,
+        isEmpty: false,
+        isError: false
+      };
+    } else {
+      this.state = {
+        ...this.state,
+        isError: props.isError || props.isRequired || false
+      };
+    }
+  }
+
+  componentWillReceiveProps ({ contractAbi, isError, isRequired }: Props) {
+    if (contractAbi) {
```

Comment:
Same comment as above. Would probably suggest allow `null` for contractAbi, so you can do `contractAbi: contractAbi || null` in the state.